tests/test_client.py

```diff
@@ -0,0 +1,315 @@
+            self.base_url = 'https://api.atlassian.com/ex/jira/{}{}'
+            self.cloud_id = config.get('cloud_id')
+            self.access_token = config.get('access_token')
+            self.refresh_token = config.get('refresh_token')
+            self.oauth_client_id = config.get('oauth_client_id')
+            self.oauth_client_secret = config.get('oauth_client_secret')
+
+            # Only appears to be needed once for any 6 hour period. If
+            # running the tap for more than 6 hours is needed this will
+            # likely need to be more complicated.
+            self.refresh_credentials()
+            self.test_credentials_are_authorized()
+        else:
+            LOGGER.info("Using Basic Auth API authentication")
+            self.base_url = config.get("base_url")
+            self.auth = HTTPBasicAuth(config.get("username"), config.get("password"))
+
+    def url(self, path):
+        if self.is_cloud:
+            return self.base_url.format(self.cloud_id, path)
+
+        # defend against if the base_url does or does not provide https://
+        base_url = self.base_url
+        base_url = re.sub('^http[s]?://', '', base_url)
+        base_url = 'https://' + base_url
+        return base_url.rstrip("/") + "/" + path.lstrip("/")
+
+    def _headers(self, headers):
+        headers = headers.copy()
+        if self.user_agent:
+            headers["User-Agent"] = self.user_agent
+
+        if self.is_cloud:
+            # Add OAuth Headers
+            headers['Accept'] = 'application/json'
+            headers['Authorization'] = 'Bearer {}'.format(self.access_token)
+
+        return headers
+
+    @backoff.on_exception(backoff.expo,
+                          (requests.exceptions.ConnectionError, HTTPError),
+                          jitter=None,
+                          max_tries=6,
+                          giveup=lambda e: not should_retry_httperror(e))
+    def send(self, method, path, headers={}, **kwargs):
+        if self.is_cloud:
+            # OAuth Path
+            request = requests.Request(method,
+                                       self.url(path),
+                                       headers=self._headers(headers),
+                                       **kwargs)
+        else:
+            # Basic Auth Path
+            request = requests.Request(method,
+                                       self.url(path),
+                                       auth=self.auth,
+                                       headers=self._headers(headers),
+                                       **kwargs)
+        return self.session.send(request.prepare())
+
+    @backoff.on_exception(backoff.constant,
+                          RateLimitException,
+                          max_tries=10,
+                          interval=60)
+    def request(self, tap_stream_id, *args, **kwargs):
+        wait = (self.next_request_at - datetime.now()).total_seconds()
+        if wait > 0:
+            time.sleep(wait)
```

Comment:
Is this sleep necessary?

Reply:
So this was copied from the http.py file that is in the tap. It looks like it was added in 2017 by Buck Ryan, someone who is no longer on the team or it was open source? Anyway, I saw this at the top if you look at the constant used that creates the wait time:
```
# The project plan for this tap specified:
# > our past experience has shown that issuing queries no more than once every
# > 10ms can help avoid performance issues
TIME_BETWEEN_REQUESTS = timedelta(microseconds=10e3)
```
In my personal opinion this is why backoff is supposed to add incremental backoff with jitter, not add wait time like this... I mean I can test without it but I'd rather not spend a lot of time fiddling with something that works either. @dmosorast @cosimon @luandy64 thoughts?

Reply:
Discussed in Slack, but very likely not for this application. It was a gut feeling spacer from an earlier rendition of this integration.

Reply:
Removed in the latest commit. Again I only changed the test_client. I didn't remove this from the tap itself, even if I suspect we might want to do that I think this PR is big enough.